[Inductor-FX] Don't flatten constant args

chenmillie opened the pull request:
Summary: Fallback kernels are created with flattened constant args and an `unflatten` utility to unflatten them when needed. Apply it in FXConverter to preserve the original structure

Test Plan: added new CI tests

Differential Revision: D85347589




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

blaine-rister reviewed (approved):
LGTM! I left one comment; please see that before merging.

blaine-rister commented on torch/_inductor/codegen/wrapper_fxir.py:
```diff
@@ -1113,11 +1113,15 @@ def _generate_extern_kernel_common(
         assert ir.is_node_sequence(kernel.inputs)
         tensor_nodes = tuple(self._generate_buffer(arg) for arg in kernel.inputs)
-        args = tensor_nodes + tuple(kernel.constant_args)
+        if hasattr(kernel, "unflatten_args"):
```
Have we seen any cases where `kernel` doesn't have this attr? Since `kernel` has type `ir.ExternKernel`, it seems like it should always have an `unflatten_args` method. Then again, Python type annotations can be misleading.

chenmillie replied:
Sometimes it can be an `ir.ExternKernelOut`, which doesn't have the attribute.

chenmillie committed:
[Inductor-FX] Don't flatten constant args (#166144)

Summary:
Fallback kernels are created with flattened constant args and an `unflatten` utility to unflatten them when needed. Apply it in FXConverter to preserve the original structure


Test Plan:
added new CI tests





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

Differential Revision: D85347589

Pulled By: chenmillie
